### ISSUE TITLE
Filter by arbitrary callback function

### ISSUE THREAD

**Issue by nolar** [zalando-incubator/kopf#98]
In some cases, it might be needed that only a subset of resources is served and handled. Some of such filtering are suggested in #45 #58, but it can be anything arbitrary.

Without such filtering, the logs will be cluttered by the resource events/causes that are of no interest to the operator.

For this, a callback function can be specified on the handler declaration:

```python
KNOWN_PODS = {}

@kopf.on.event('', 'v1', 'pods'
               when=lambda event, uid, **_:  event=='ADDED' and uid not in KNOWN_PODS)
def new_pod_created(**kwargs):
    pass
```

The callback should be called via the same invocation protocol as the handlers (see `kopf.reactor.invocation`), i.e. with all the handler's kwargs — be that an event-handler or a cause-handler (different sets of kwargs).

If the callback returns true, then the object is handled as normally. When it returns false, the handler is ignored, maybe the whole event is ignored (if there are no other handlers) — and therefore **produces no single line of logs at all**.

No additional kwargs are injected for such filtering.

----

Related: 

* #30 for silent handlers (no status/progress storage).
* #84 for pods support (but can be tested with another custom resource).
* #58 for filtering by owner references (aka parents).
* #45 for filtering by labels/annotations.

Checklist:

* [x] Acceptance criteria:
  * [x] Event handler registered with the filtering callback function/lambda/partial.
  * [x] Callback is invoked every time an event is arrived and/or cause is detected.
  * [x] Events/handlers are filtered by the result of this callback invocation.
  * [x] No logs about handler or event/cause are produced is the handler was filtered out.
  * [x] Minimum overhead and not extra calls if the callback is not specified (the default).
* [x] Tests.
* [x] Documentation.

**Comment by Jc2k** [comment]
This would be really handy for me.

I was thinking about writing a cert-manager plugin with kopf - something like [this](https://github.com/jetstack/cert-manager-external-issuer-example).

With this feature I would be able to filter out `CertificateRequests` that weren't "assigned" to my operator by looking at `spec.issuerRef`.

(Presumably these objects would then not be tagged with the kopf last-configuration annotations etc too).

**Comment by nolar** [comment]
I think, it is easily implementable now. Since at least some filters were added in #123 (by labels/annotations), adding more filters is easy. 

A starting place to look for: https://github.com/zalando-incubator/kopf/blob/0.23.2/kopf/reactor/registries.py#L570-L623

**Comment by Jc2k** [comment]
Thanks for the hint. I've had a bash at this in #258.

**Comment by nolar** [comment]
Released as [`kopf==0.25`](https://github.com/zalando-incubator/kopf/releases/tag/0.25). 

@Jc2k Thank you very much for your contribution!
